### PR TITLE
API Wrapper Overhaul 

### PR DESCRIPTION
**Pull request by THoisington**
I'm not sure if any of the original code remains after the following changes were made: 

- Converted to class-based API
- Covered all endpoints for Get Requests, including those that were missing
- Added some "convenience" methods 
- Wrote integration tests 
- Integrated Axios, Moment, QS, Deep-Map-Keys, and CamelCase 
- First update to README, but just as an example.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,24 +7,45 @@ The Node/JavaScript Client for the [Clockodo-API](https://www.clockodo.com/de/ap
 Get user (email) and clockodo api key on "[My area](https://my.clockodo.com/en/users/editself)".
```
[the remark this answers is not in the document]
Comment (THoisington) [reply README.md]: Maybe it would be worth keeping the initial README simple like that and linking to the Github Wiki for the API write-up?

File: lib/api.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    getClockRunning() {
+        return this[clockodoApi].get("/clock");
+    }
+
+    getClockStatus() {
+        return this[clockodoApi].get("/clock/update");
+    }
+
+    // Convienece Methods
```
[the remark this answers is not in the document]
Comment (THoisington) [reply lib/api.js]: I'm totally with you on this.

File: lib/api.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    getServiceEntriesBetween(serviceId, begin, end) {
+        return this.getEntries(begin, end, { filterServiceId: serviceId });
+    }
+    getProjectEntriesBetween(projectId, begin, end) {
+        return this.getEntries(begin, end, { filterProjectId: projectId });
+    }
+}
+
+function _checkRequired(params, requiredList) {
```
[the remark this answers is not in the document]
Comment (THoisington) [reply lib/api.js]: I think for now I am going to stick with it and think about it. I was against them at first but like many things in JS my opinion is constantly evolving haha.

File: tests/integration/api.integration.test.js [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                    active: true,
+                    editLock: null,
+                },
+            ],
+        };
+
+        expect.assertions(1);
+
+        return clockodo.getUsers().then(data => {
+            expect(Object.keys(data.users[0]).sort()).toEqual(Object.keys(mockResponse.users[0]).sort());
```
[the remark this answers is not in the document]
Comment (THoisington) [reply tests/integration/api.integration.test.js]: Well then that is comparing values where I just wanted to compare the object structure (and ensure data was actually coming back at the same time). I think we will need to have a sit down and look at the tests after I add some 😅

File: lib/api.js [hunk 5]
```diff
@@ -28,15 +28,25 @@ class Clockodo {
             throw new Error("No Clockodo 'apiKey' set");
         }
 
-        this[clockodoApi] = new ClockodoApi(user, apiKey);
+        this[clockodoApi] = new ClockodoLib({ user, apiKey });
```
Comment (jhnns) [inline lib/api.js]: I'm fine with renaming `ClockodoApi` to `ClockodoLib`. For consistency reasons, you should also rename the symbol to `clockodoLib`.

File: .eslintrc.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
+  "extends": ["peerigon", "peerigon/node"],
+  "root": true,
+  "rules": {
+    "camelcase": "off"
```
Comment (jhnns) [inline .eslintrc.json]: I don't think it's necessary to disable the rule globally. You can disable it locally like:

```js
const params = {
    // API params are in snake case
    /* eslint-disable camelcase */
    some_param: true,
    other_param: true,
    /* eslint-enable camelcase */
}
```

It's also good to leave a comment why you've disabled the rule.

You can disable the rules for the tests though (in `tests/.eslintrc.json`).

File: lib/api.js [hunk 7]
```diff
@@ -28,15 +28,25 @@ class Clockodo {
...
 
-    getUsers() {
-        return this[clockodoApi].get("/users");
+    getAbsences(params) {
+        _checkRequired(params, REQUIRED_PARAMS_GET_ABSENCES);
+
+        return this[clockodoApi].get("/absences", params);
+    }
+
+    getClockRunning() {
```
Comment (jhnns) [inline lib/api.js]: Personally, I would stick to the API endpoint naming as close as possible to avoid confusion (unless there's a good cause). So this should be `getClock()`.

File: lib/api.js [hunk 8]
```diff
@@ -28,15 +28,25 @@ class Clockodo {
...
+        return this[clockodoApi].get("/clock");
+    }
+
+    getClockStatus() {
```
Comment (jhnns) [inline lib/api.js]: Likewise, this should be `getClockUpdate()`

File: lib/api.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+const REQUIRED_PARAMS_GET_TASK_DURATION = [
+    "taskCustomerId",
+    "taskProjectId",
+    "taskServiceId",
+    "taskText",
+    "taskBillable",
+];
+const REQUIRED_PARAMS_GET_ENTRIES = ["begin", "end"];
+const REQUIRED_PARAMS_GET_ENTRYGROUPS = ["begin", "end", "grouping"];
+const REQUIRED_PARAMS_GET_USERREPORTS = ["year"];
```
Comment (jhnns) [inline lib/api.js]: Since these names relate to the public camelcase methods, they should be:
- `REQUIRED_PARAMS_GET_USERREPORTS` -> `REQUIRED_PARAMS_GET_USER_REPORTS`
- `REQUIRED_PARAMS_GET_USERREPORTS` -> `REQUIRED_PARAMS_GET_USER_REPORTS`

File: lib/lib.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
+"use strict";
+
+const axios = require("axios");
+const camelCase = require("camelcase");
+const deepMapKeys = require("deep-map-keys");
+const qs = require("qs");
+
+const ENDPOINT = "https://my.clockodo.com/api";
+const axiosClient = Symbol("axiosClient");
+const paramMapping = {
```
Comment (jhnns) [inline lib/lib.js]: I guess it makes sense to move that param mapping object to dedicated file. Then you can just link to that file in the `README` if someone wants to look up the param name.

File: package.json [hunk 11]
```diff
@@ -1,32 +1,43 @@
...
+    "license": "Unlicense",
+    "bugs": {
+        "url": "https://github.com/peerigon/clockodo-node/issues"
+    },
+    "homepage": "https://github.com/peerigon/clockodo-node#readme",
+    "dependencies": {
+        "axios": "^0.18.0",
+        "camelcase": "^5.0.0",
+        "deep-map-keys": "^1.2.0",
+        "dotenv": "^5.0.1",
```
Comment (jhnns) [inline package.json]: I think `dotenv` is a dev dependency?

File: package.json [hunk 12]
```diff
@@ -1,32 +1,43 @@
...
+        "moment": "^2.21.0",
+        "node-rest-client": "^3.1.0",
```
Comment (jhnns) [inline package.json]: `node-rest-client` isn't necessary anymore

File: package.json [hunk 13]
```diff
@@ -1,32 +1,43 @@
...
+        "qs": "^6.5.1",
+        "when": "^3.7.8"
```
Comment (jhnns) [inline package.json]: `when` isn't necessary too

File: package.json [hunk 14]
```diff
@@ -1,32 +1,43 @@
...
+    },
+    "devDependencies": {
+        "eslint": "^4.18.2",
+        "eslint-config-peerigon": "^12.0.1",
+        "eslint-plugin-import": "^2.9.0",
+        "eslint-plugin-jsdoc": "^3.5.0",
+        "jest": "^22.4.2",
+        "nock": "^9.2.3",
+        "prompt": "^1.0.0"
```
Comment (jhnns) [inline package.json]: I think it's better to have the examples as separate files. Then you don't need the `prompt` library (though I really liked your small test helper 😁)

File: tests/integration/integration.test.js [hunk 15]
```diff
@@ -0,0 +1,255 @@
+"use strict";
+
+require("dotenv").config();
+
+const Clockodo = require("../../lib/api");
+const { ClockodoLib } = require("../../lib/lib");
+
+describe("Api.js", () => {
```
Comment (jhnns) [inline tests/integration/integration.test.js]: We should skip these tests if the credentials are not provided:

```js
// These tests depend on our real Clockodo account.
// They can only be executed by Peerigon members or Travis CI.
const hasCredentials = Boolean(process.env.CLOCKODO_USER && process.env.CLOCKODO_API_KEY);

(hasCredentials ? describe : describe.skip)("Api.js", () => {
```

File: tests/integration/integration.test.js [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+    let userClockodo;
+
+    beforeEach(() => {
+        userClockodo = new Clockodo(process.env.CLOCKODO_USER, process.env.CLOCKODO_API_KEY);
+    });
+
+    describe("getUsers()", () => {
+        it("returns expected data format", () => {
+            const mockResponse = {
```
Comment (jhnns) [inline tests/integration/integration.test.js]: That's not a `mockResponse`, is it?

File: tests/integration/integration.test.js [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+                users: [
+                    {
+                        id: 2323,
+                        name: "Harry Potter",
+                        number: "7",
+                        email: "boywholived@hogwarts.com",
+                        role: "wizard",
+                        active: true,
+                        editLock: null,
```
Comment (jhnns) [inline tests/integration/integration.test.js]: Where is that user defined?

File: tests/integration/integration.test.js (hunk 15 above)
Comment (jhnns) [reply tests/integration/integration.test.js]: Btw: I'd use the class name in the describe block. So that should be `describe("Clockodo", ...`

File: tests/unit/unit.test.js [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+        }
+
+        it("correctly builds getAbsence() request", async () => {
+            expect.assertions(1);
+
+            const nockObj = getNock("/absences/007");
+
+            await userClockodo.getAbsence("007");
+
+            expect(nockObj.isDone()).toBe(true);
```
Comment (jhnns) [inline tests/unit/unit.test.js]: We can get these lines shorter if we just call `nockObj.done()`. This will throw anyway if not all interceptors have been called.

File: tests/integration/integration.test.js [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+                    },
+                ],
+            };
+
+            expect.assertions(1);
+
+            return userClockodo.getUsers().then(data => {
+                expect(Object.keys(data.users[0]).sort()).toEqual(Object.keys(mockResponse.users[0]).sort());
+            });
```
Comment (jhnns) [inline tests/integration/integration.test.js]: I think you should also use `async/await` in the tests. We will need to make it compatible with node 6 anyway, but let's do that with a separate PR.

Comment (jhnns) [reply tests/integration/integration.test.js]: I think the test for `getUsers()` should check for the `length` and the shape of the object (but not the values), like:

```js
expect(users).toHaveLength(2);
users.forEach(user => {
    expect(user).toHaveProperty("id");
    expect(user).toHaveProperty("name");
    // and so on
});
```

File: tests/integration/integration.test.js [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+                    begin: "2017-08-18 00:00:00",
+                    end: "2018-02-09 00:00:00",
+                    filterBillable: 2,
+                    filterUserId: 38557,
+                };
+
+                expect.assertions(1);
+
+                return userClockodo.getEntries(parameters).then(data => {
+                    expect(Object.keys(data.entries[0]).sort()).toEqual(expectedKeys.sort());
```
Comment (jhnns) [inline tests/integration/integration.test.js]: See comment at `getUsers()`

File: tests/integration/integration.test.js [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+                const parameters = {
+                    begin: "2017-08-18 00:00:00",
+                    end: "2018-02-09 00:00:00",
+                    grouping: ["customers_id"],
+                };
+
+                expect.assertions(1);
+
+                return userClockodo.getEntryGroups(parameters).then(data => {
+                    expect(Object.keys(data.groups[0]).sort()).toEqual(expectedKeys.sort());
```
Comment (jhnns) [inline tests/integration/integration.test.js]: See comment at `getUsers()`

File: tests/integration/integration.test.js [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+                const parameters = {
+                    begin: "2017-08-18 00:00:00",
+                    end: "2018-02-09 00:00:00",
+                    grouping: ["customers_id", "projects_id"],
+                };
+
+                expect.assertions(1);
+
+                return userClockodo.getEntryGroups(parameters).then(data => {
+                    expect(Object.keys(data.groups[0]).sort()).toEqual(expectedKeys.sort());
```
Comment (jhnns) [inline tests/integration/integration.test.js]: See comment at `getUsers()`

File: tests/integration/integration.test.js [hunk 23]
```diff
@@ -0,0 +1,255 @@
...
+        it(
+            "throws an error with missing required parameter",
+            async () => {
+                const parameters = {
+                    begin: "2017-08-18 00:00:00",
+                    filterBillable: 2,
+                    filterUserId: 38557,
+                };
+                const missingParamName = "end";
+                const expectedError = new Error(`Missing required parameter "${ missingParamName }"`);
```
Comment (jhnns) [inline tests/integration/integration.test.js]: Why not inlining `missingParamName` into the error message? Make the test as explicit and static as possible.

File: tests/integration/integration.test.js [hunk 24]
```diff
@@ -0,0 +1,255 @@
...
+        );
+        it(
+            "throws an error with missing required parameter",
+            async () => {
+                const parameters = {
+                    begin: "2017-08-18 00:00:00",
+                    end: "2018-02-09 00:00:00",
+                };
+                const missingParamName = "grouping";
+                const expectedError = new Error(`Missing required parameter "${ missingParamName }"`);
```
Comment (jhnns) [inline tests/integration/integration.test.js]: See comment about `missingParamName`

File: tests/integration/integration.test.js [hunk 25]
```diff
@@ -0,0 +1,255 @@
...
+        const goodResource = "users";
+
+        expect.assertions(1);
+
+        return expect(libClockodo.get(goodResource)).resolves.toHaveProperty("users");
+    });
+
+    it("returns with immediate access to data", () => {
+        const getUser = "/users/" + process.env.CLOCKODO_TEST_ID;
+        const expectedName = process.env.CLOCKODO_TEST_NAME;
```
Comment (jhnns) [inline tests/integration/integration.test.js]: These env variables where missing in my `.env`. Can you create an `.env.example` with all the env variable names.

File: tests/unit/unit.test.js [hunk 26]
```diff
@@ -0,0 +1,222 @@
...
+
+const { stringify } = require("querystring");
+const qs = require("qs");
+const nock = require("nock");
+const Clockodo = require("../../lib/api");
+
+describe("Api.js", () => {
+    const userClockodo = new Clockodo("test", "test");
+
+    describe("HTTP Request For All API Methods", () => {
```
Comment (jhnns) [inline tests/unit/unit.test.js]: This `describe()` block is unnecessary (it's the only one on that level) and I think it's not actually describing your test.

Usually, the `describe()` block contains the *subject* of the `it()` blocks.

So, in your case, I would write:

```js
describe("Clockodo (instance)", () => {
    const userClockodo = new Clockodo("test", "test");

    describe("getAbsences()", () => {
        it("triggers a GET request to /absences", () => {});
        it("throws when a required parameter is missing", () => {});
    });


    describe("getCustomer(id)", () => {
        it("triggers a GET request to /customers with the given id", () => {});
        it("throws when a required parameter is missing", () => {});
    });
```

File: tests/integration/integration.test.js [hunk 27]
```diff
@@ -0,0 +1,255 @@
...
+                expect.assertions(1);
+
+                return userClockodo.getEntries(parameters).then(data => {
+                    expect(Object.keys(data.entries[0]).sort()).toEqual(expectedKeys.sort());
+                });
+            },
+            10000
+        );
+        it(
+            "throws an error with missing required parameter",
```
Comment (jhnns) [inline tests/integration/integration.test.js]: I think, this test should be part of our unit tests. Remember that the integration tests can only be executed by developers that have our credentials. They should only run a few basic tests on the real API.

File: tests/unit/unit.test.js [hunk 28]
```diff
@@ -0,0 +1,222 @@
...
+        function getNock(path) {
+            return nock("https://my.clockodo.com/api")
+                .get(path)
+                .reply(200);
+        }
```
Comment (jhnns) [inline tests/unit/unit.test.js]: I think I wouldn't extract this as a helper function. I like it when the nock configuration is repeated for every test because they are basically like an assertion (at least in combination with `scope.done()`). That's why it's good to have them close to the actual test. I also think that it might get more complicated as soon as we support other HTTP verbs or need to handle different HTTP status codes.

File: lib/api.js [hunk 29]
```diff
@@ -0,0 +1,161 @@
...
+const REQUIRED_PARAMS_GET_ABSENCES = ["year"];
+
+class Clockodo {
+    constructor(user, apiKey) {
```
[the remark this answers is not in the document]
Comment (THoisington) [reply lib/api.js]: By the way I think I did this the way you wanted?

File: lib/api.js (hunk 7 above)
Comment (THoisington) [reply lib/api.js]: Sure thing, I just didn't think their choices made much sense

File: lib/lib.js (hunk 10 above)
Comment (THoisington) [reply lib/lib.js]: I would be fine moving it to a separate file, but when I was reading your comments about possibly leaving out all mention of us changing the parameters I thought that might be best because the user shouldn't have to worry about that. I would honestly be happy if they didn't have to think about the Clockodo docs at all. But we can discuss this in person. I will for now go ahead and move it out to a separate file.

File: tests/integration/integration.test.js (hunk 19 above)
Comment (THoisington) [reply tests/integration/integration.test.js]: Let's take a look at the different ways we can write this test later

File: tests/integration/integration.test.js [hunk 30]
```diff
@@ -3,35 +3,26 @@
...
-                ],
-            };
+            const expectedKeys = ["id", "name", "number", "email", "role", "active", "editLock"];
 
             expect.assertions(1);
 
-            return userClockodo.getUsers().then(data => {
-                expect(Object.keys(data.users[0]).sort()).toEqual(Object.keys(mockResponse.users[0]).sort());
+            return clockodo.getUsers().then(data => {
+                expect(Object.keys(data.users[0]).sort()).toEqual(expectedKeys.sort());
```
Comment (jhnns) [inline tests/integration/integration.test.js]: Can you refactor these tests to `async/await`? 😁

File: tests/unit/unit.test.js [hunk 31]
```diff
@@ -5,82 +5,91 @@ const qs = require("qs");
...
         });
-
-        it("correctly builds getAbsences() request", async () => {
+        it("throws an error when getAbsences() is missing param", async () => {
             expect.assertions(1);
 
+            try {
+                await clockodo.getAbsences({});
+            } catch (e) {
+                expect(e).toEqual(Error('Missing required parameter "year"'));
```
Comment (jhnns) [inline tests/unit/unit.test.js]: It's really uncommon to call `Error` without `new`. I'd rather check for the error message instead since you're comparing error objects that are actually not equal (the `.stack` property differs, but Jest seems to handle that).

I think that the assertion error would also became more readable if you only compare the `.message`.

File: tests/unit/unit.test.js [hunk 32]
```diff
@@ -187,36 +231,116 @@ describe("Api.js", () => {
...
+            }
+        });
+    });
+
+    describe("Clockodo Constructor", () => {
+        /* eslint-disable no-unused-vars */
+        it("throws an error when constructor is missing user email", () => {
+            try {
+                const clockodoMissingUser = new Clockodo({ user: undefined, apiKey: "dfdsg34t643" });
```
Comment (jhnns) [inline tests/unit/unit.test.js]: You can write `void new Clockodo(...` to get rid of the unnecessary ESLint errors now. You can also remove the disable rule comment then.

I have to update [`peerigon/tests`](https://github.com/peerigon/eslint-config-peerigon/blob/master/tests.js) because it's very common to call `new SomeConstructor` while not using the result in these tests.
